spark_log_parser/loaders/json.py

```diff
@@ -13,53 +13,72 @@
+class JSONLinesDataLoader(DataLoader):
     cache = False
     lines_data_loader: RawJSONLinesDataLoader
 
     @staticmethod
-    def yield_json_lines(filename, lines) -> Iterator[str]:
-        num_bad_lines_seen = 0
-        for line in lines:
+    def _yield_json_lines(data):
+        for filepath, file in data:
+            logger.info(f"Processing: {filepath}")
+
+            lines = iter(file)
+            first_line = next(lines)
             try:
-                yield orjson.loads(line)
-            # Note - this is largely here because we may get arbitrary file types in archives where we are
-            #  expecting eventlog files (which are JSON Lines). If we try to load those "lines" as
-            #  JSON objects, they will fail, and so this allows us to just skip those "bad" lines and
-            #  continue processing other files in the archive. However, this `except` could maybe be
-            #  more robust so that we aren't ignoring real issues! If we do encounter some bad lines,
-            #  we will at the very least log how many we hit
+                json_line = orjson.loads(first_line)
+                yield json_line
             except orjson.JSONDecodeError:
-                num_bad_lines_seen += 1
-                # We have to continue here because these will be the lines for the top-level file. If this is an
-                #  archive, then it is possible for some lines to not be parse-able as JSON, but for others to be
-                continue
-
-        if num_bad_lines_seen > 0:
-            logger.info(f"Failed to parse {num_bad_lines_seen} JSON lines in top-level file: {filename}")
+                # If the first line is not parseable as a JSON object, try to parse the whole file as a JSON Object
+                #  as well, and yield that as a line instead.
+                # We do this for a few reasons -
+                #  - Sometimes regular JSON files are delivered in archives along with eventlog files (which are JSON
+                #     Lines). These JSON files may contain valuable information that we don't want to drop on the floor
+                #     (for example, we may receive Databricks pricing information, which we may be able to use!)
+                #  - We may not be able to tell without opening the file whether the eventlog provided to us is a "raw"
+                #     eventlog (i.e. delivered to us in exactly the way Spark output them), or if it is an
+                #     already-parsed log.
+                #
+                # This means that, in a sense, we treat regular JSON files as a special case of JSON Lines files
+                for line in lines:
+                    first_line += line
+
+                try:
+                    yield orjson.loads(first_line)
+                except orjson.JSONDecodeError:
+                    logger.warning(f"Could not parse file {filepath} as JSON - skipping")
+                finally:
+                    continue
```

Comment:
As mentioned in the PR description, I'm also 100% open to just skipping all of the lines in this file instead of reading them all into 1 object. The tricky thing is that we don't know before attempting to load a file whether or not it is a JSON Lines or just a regular JSON file, which means that we don't have a good way of handling parsed logs without opening it up first. I can think of a few ways to handle that, but thought I'd propose this first as it seems to me to be the simplest solution.